### PR TITLE
[MYST-41] Session generation and exchange

### PR DESCRIPTION
**Pull request by ignasbernotas**

### CODE REVIEW

File: cmd/mysterium_client/command_run/command.go [hunk 1]
```diff
@@ -38,13 +40,20 @@ func (cmd *CommandRun) Run(options CommandOptions) (err error) {
...
 	if err != nil {
 		return err
 	}
 
+	vpnSession, err := getVpnSession(vpnSessionJson)
+	if err != nil {
+		return err
+	}
+
+	session.Id = string(vpnSession.Id)
```
Comment (Waldz) [inline cmd/mysterium_client/command_run/command.go]: Add todo for removing this?

File: cmd/mysterium_client/command_run/command.go [hunk 2]
```diff
@@ -38,13 +40,20 @@ func (cmd *CommandRun) Run(options CommandOptions) (err error) {
 		return err
 	}
 
-	vpnConfigString, err := sender.Request(communication.GET_CONNECTION_CONFIG, "")
+	vpnSessionJson, err := sender.Request(communication.SESSION_CREATE, string(proposal.Id))
 	if err != nil {
 		return err
 	}
 
+	vpnSession, err := getVpnSession(vpnSessionJson)
```
Comment (Waldz) [inline cmd/mysterium_client/command_run/command.go]: `getVpnSession()` could be a wrapper: make request + parse response

Comment (Waldz) [reply cmd/mysterium_client/command_run/command.go]: I am missing test, which freezes response of `SESSION_CREATE `

File: session/generate.go [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+
+import (
+	"crypto/rand"
+	"encoding/hex"
+)
+
+type SessionId string
+
+func GenerateSessionId() (sid SessionId, err error) {
+	b, err := generateRandomBytes(16)
```
Comment (Waldz) [inline session/generate.go]: Did You consider using UniqueID library, which has randomness issues solved?
I used https://github.com/satori/go.uuid:
```
requestId := uuid.NewV4().String()
```

File: session/manager.go [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+	Sessions []SessionId
+}
+
+func (manager *Manager) Create() (SessionId, error) {
+	sid, err := GenerateSessionId()
+	if err != nil {
+		return sid, err
+	}
+
+	for !manager.sessionIsUnique(sid) {
```
Comment (Waldz) [inline session/manager.go]: What if SID will have collision with another node's session?
Maybe let add ticket for future improvement

File: session/manager_test.go [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+import (
+	"testing"
+	"github.com/stretchr/testify/assert"
+)
+
+func TestManagerHasUniqueSessionsStored(t *testing.T) {
+	manager := Manager{}
+	length := 10
+
+	for i := 0; i < length; i++ {
```
Comment (Waldz) [inline session/manager_test.go]: Looks like loop with overhead. 
You can:
- create SID
- test private function `sessionIsUnique()`

File: openvpn/session/factory.go [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+import (
+	node_session "github.com/mysterium/node/session"
+)
+
+var manager node_session.Manager
+
+func NewVpnSession(config string) (session VpnSession, err error) {
+	id, err := manager.Create()
+	if err != nil {
+		return VpnSession{}, err
```
Comment (Waldz) [inline openvpn/session/factory.go]: You can just return:
```
return
```
or
```
return session, err
```

File: openvpn/session/factory.go [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+	}
+
+	vpnSession := VpnSession{
```
Comment (Waldz) [inline openvpn/session/factory.go]:
```
session = VpnSession{...}
```

File: communication/message_types.go [hunk 8]
```diff
@@ -2,6 +2,6 @@ package communication
 
 const (
 	// Client is trying to establish new dialog with Node
-	DIALOG_CREATE         = RequestType("dialog-create")
-	GET_CONNECTION_CONFIG = RequestType("get-connection-config")
+	DIALOG_CREATE  = RequestType("dialog-create")
+	SESSION_CREATE = RequestType("session-create")
```
Comment (shroomist) [inline communication/message_types.go]: I suggest defining structs for message types
```golang
type SessionCreate struct {
proposalId, serviceStartParams
}
```
we also need response types defined here

Comment (ignasbernotas) [reply communication/message_types.go]: Message types do have their own structs:) These constants are basically topics for the communication channel, so we still need them.

File: cmd/mysterium_client/command_run/command.go [hunk 9]
```diff
@@ -38,13 +42,13 @@ func (cmd *CommandRun) Run(options CommandOptions) (err error) {
 		return err
 	}
 
-	vpnConfigString, err := sender.Request(communication.GET_CONNECTION_CONFIG, "")
+	vpnSession, err := getVpnSession(sender, strconv.Itoa(proposal.Id))
```
Comment (Waldz) [inline cmd/mysterium_client/command_run/command.go]: Why casting in controller level

File: cmd/mysterium_client/command_run/command.go [hunk 10]
```diff
@@ -75,3 +79,23 @@ func (cmd *CommandRun) Kill() {
...
+
+	var response vpn_session.SessionCreateResponse
+
+	err = json.Unmarshal([]byte(sessionJson), &response)
+	if err != nil {
+		return
+	}
+
+	if response.Success == false {
+		return session, errors.New(response.Message)
```
Comment (Waldz) [inline cmd/mysterium_client/command_run/command.go]: Error might return empty. I suggest adding prefix string

File: cmd/mysterium_server/command_run/command.go [hunk 11]
```diff
@@ -45,11 +47,28 @@ func (cmd *CommandRun) Run(options CommandOptions) (err error) {
...
+				options.DirectoryConfig+"/ta.key",
+			)
+		},
+	}
 	handleDialog := func(sender communication.Sender, receiver communication.Receiver) {
-		receiver.Respond(communication.GET_CONNECTION_CONFIG, func(request string) (response string) {
-			config, _ := buildVpnClientConfig(vpnServerIp, options.DirectoryConfig)
-			return config
-		})
+		receiver.Respond(communication.SESSION_CREATE, sessionResponseHandler.Handle)
```
Comment (Waldz) [inline cmd/mysterium_server/command_run/command.go]: spacy spaces

File: openvpn/session/create_response.go [hunk 12]
```diff
@@ -0,0 +1,68 @@
...
+	}
+
+	if scr.ProposalId != numericProposalId {
+		response.Success = false
+		response.Message = "Proposal doesn't exist."
+
+		return serializeCreateResponse(response)
+	}
+
+	str := serializeCreateResponse(scr.buildResponse())
```
Comment (Waldz) [inline openvpn/session/create_response.go]: Nice segragation

File: openvpn/session/create_response_test.go [hunk 13]
```diff
@@ -0,0 +1,90 @@
...
+				"id": "",
+				"config": ""
+			}
+		}`,
+		sessionResponse,
+	)
+}
+
+func TestHandler_Success(t *testing.T) {
+	handler := CreateResponseHandler{
```
Comment (Waldz) [inline openvpn/session/create_response_test.go]: You could reuse same handler|code-block in all tests

File: session/manager_test.go [hunk 14]
```diff
@@ -0,0 +1,23 @@
+package session
+
+import (
+	"github.com/stretchr/testify/assert"
+	"testing"
+)
+
+func TestManagerHasSessionsStored(t *testing.T) {
```
Comment (Waldz) [inline session/manager_test.go]: Does not look like unit.
Is it like integration test?

File: cmd/mysterium_server/command_run/command.go (hunk 11 above)
Comment (ignasbernotas) [reply cmd/mysterium_server/command_run/command.go]: It seems so on github, IDE shows it fine.

File: session/manager_test.go (hunk 14 above)
Comment (ignasbernotas) [reply session/manager_test.go]: Yeah, testing the manager is kind of pointless in this case.. the only use case for unit tests would be to freeze the interface methods, does that make sense though?

File: session/manager.go [hunk 15]
```diff
@@ -0,0 +1,27 @@
...
+}
+
+func (manager *Manager) Create() SessionId {
+	id := manager.Generator.Generate()
+	manager.Add(id)
+
+	return id
+}
+
+func (manager *Manager) Get(sid SessionId) (sessionId SessionId) {
```
Comment (Waldz) [inline session/manager.go]: Impossible to distinguish if session was found/not-found.
Test would help You to prove usability.

File: cmd/mysterium_client/command_run/command.go (hunk 10 above)
Comment (ignasbernotas) [reply cmd/mysterium_client/command_run/command.go]: Don't see how that could happen. All edge cases are handled. What did I miss?:)

Comment (Waldz) [reply cmd/mysterium_client/command_run/command.go]: Because it's external message, can come from another node's code

**Pull request merged**